File: Python/string_algorithms/01_trie_construction.py
import sys
from collections import defaultdict
from itertools import count


class Trie(object):
    def __init__(self, words=None):
        self._node_counter = count(0)
        self._root = next(self._node_counter)
        self._trie = defaultdict(list)
        if words is not None:
            self.add_words_to_trie(words)

    @property
    def trie(self):
        return dict(self._trie)

    def add_words_to_trie(self, words):
        for word in words:
            self._add_to_trie(word)

    def _add_to_trie(self, word):
        current_node = self._root
        for letter in word:
            node = self._get_neighbour_with_given_label(current_node, letter)
            if node is not None:
                current_node = node
            else:
                next_node = next(self._node_counter)
                self._trie[current_node].append((next_node, letter))
                current_node = next_node

    def _get_neighbour_with_given_label(self, current_node, letter):
        neighbours = self._trie[current_node]
        for node, label in neighbours:
            if label == letter:
                return node


def print_trie_adjacency_list(trie):
    for node, neighbours in trie.items():
        for neighbour, label in neighbours:
            print("{}->{}:{}".format(node, neighbour, label))


def main():
    data = sys.stdin.read().splitlines()
    patterns = data[1:]
    trie = Trie(patterns)
    print_trie_adjacency_list(trie.trie)


if __name__ == "__main__":
    main()

File: Rust/string_algorithms/src/bin/01_trie_construction.rs
use std::collections::HashMap;
use std::io;

struct Trie {
    adjacency_list: HashMap<i32, Vec<(i32, char)>>
}

impl Trie {
    fn build(words: &[String]) -> Self {
        let mut node_counter = 0;
        let mut adjacency_list: HashMap<i32, Vec<(i32, char)>> = HashMap::new();
        for word in words {
            let mut current_node = 0;
            for character in word.chars() {
                let node_with_character_as_label =
                    adjacency_list
                        .get(&current_node)
                        .and_then(|ns| ns.iter().find(|(_, label)| *label == character).map(|(n, _)| *n));
                match node_with_character_as_label {
                    Some(node) => current_node = node,
                    None => {
                        node_counter += 1;
                        adjacency_list.entry(current_node).or_insert_with(Vec::new).push((node_counter, character));
                        current_node = node_counter;
                    }
                };
            }
        }

        Trie { adjacency_list }
    }
}

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    line
}

fn read_patterns(nr_patterns: i32) -> Vec<String> {
    let mut patterns = Vec::with_capacity(nr_patterns as usize);
    for _ in 0..nr_patterns {
        let pattern = read_line().trim().to_string();
        patterns.push(pattern);
    }

    patterns
}

fn main() {
    let nr_patterns = read_line().trim().parse().unwrap();
    let patterns = read_patterns(nr_patterns);
    let trie = Trie::build(&patterns);
    for (node, neighbors) in trie.adjacency_list {
        for (neighbor, label) in neighbors {
            println!("{}->{}:{}", node, neighbor, label);
        }
    }
}
